backends/tc/ebpfCodeGen.cpp

```diff
@@ -0,0 +1,1481 @@
+    commentGen.setBuilder(builder);
+
+    unsigned int structAlignment = 4;  // 4 by default
+
+    builder->emitIndent();
+    builder->appendLine("u32 keysz;");
+    builder->emitIndent();
+    builder->appendLine("u32 maskid;");
+
+    if (keyGenerator != nullptr) {
+        for (auto c : keyGenerator->keyElements) {
+            auto mtdecl = program->refMap->getDeclaration(c->matchType->path, true);
+            auto matchType = mtdecl->getNode()->to<IR::Declaration_ID>();
+
+            if (!isMatchTypeSupported(matchType)) {
+                ::error(ErrorType::ERR_UNSUPPORTED, "Match of type %1% not supported",
+                        c->matchType);
+            }
+
+            auto ebpfType = ::get(keyTypes, c);
+            cstring fieldName = ::get(keyFieldNames, c);
+
+            if (ebpfType->is<EBPF::EBPFScalarType>() &&
+                ebpfType->to<EBPF::EBPFScalarType>()->alignment() > structAlignment) {
+                structAlignment = 8;
+            }
+
+            builder->emitIndent();
+            ebpfType->declare(builder, fieldName, false);
+
+            builder->append("; /* ");
+            c->expression->apply(commentGen);
+            builder->append(" */");
+            builder->newline();
+        }
+    }
+
+    // Add dummy key if P4 table define table with empty key. This due to that hash map
+    // cannot have zero-length key. See function htab_map_alloc_check in kernel/bpf/hashtab.c
+    // located in Linux kernel repository
+    if (keyFieldNames.empty()) {
+        builder->emitIndent();
+        builder->appendLine("u8 __dummy_table_key;");
+    }
+
+    builder->blockEnd(false);
+    builder->appendFormat(" __attribute__((aligned(%d)))", structAlignment);
+    builder->endOfStatement(true);
+
+    if (isTernaryTable()) {
+        // generate mask key
+        builder->emitIndent();
+        builder->appendFormat("#define MAX_%s_MASKS %u", keyTypeName.toUpper(),
+                              program->options.maxTernaryMasks);
+        builder->newline();
+
+        builder->emitIndent();
+        builder->appendFormat("struct %s_mask ", keyTypeName.c_str());
+        builder->blockStart();
+
+        builder->emitIndent();
+        builder->appendFormat("__u8 mask[sizeof(struct %s)];", keyTypeName.c_str());
+        builder->newline();
+
+        builder->blockEnd(false);
+        builder->appendFormat(" __attribute__((aligned(%d)))", structAlignment);
+        builder->endOfStatement(true);
+    }
+}
+
+void EBPFTablePNA::emitValueType(EBPF::CodeBuilder *builder) {
+    emitValueActionIDNames(builder);
+
+    // a type-safe union: a struct with a tag and an union
+    builder->emitIndent();
+    builder->appendFormat("struct __attribute__((__packed__)) %s ", valueTypeName.c_str());
+    builder->blockStart();
+
+    emitValueStructStructure(builder);
+    emitDirectValueTypes(builder);
+
+    builder->blockEnd(false);
+    builder->endOfStatement(true);
+
+    if (isTernaryTable()) {
```

Comment:
This part is not required as well - ternary matching is handled entirely in P4TC

Reply:
Removed